### PR TITLE
Github actions - update_apps.yaml workflow

### PR DESCRIPTION
**Pull request by theofficialgman**
This workflow is in beta. Currently it runs only when manually told to do so (in the future that will be automatic say once a day).

It also commits directly to the branch which it was run from. It has been suggested that this be changed to instead make a PR every time an app update is found, this could be done with a change of `EndBug/add-and-commit@v7` to one of the many PR actions from the marketplace

### CODE REVIEW

File: .github/workflows/update_apps.yml
Comment (Crilum) [inline .github/workflows/update_apps.yml:23]:
```suggestion
      - name: Run app update scripts
```

File: .github/workflows/update_apps.yml
Comment (Crilum) [inline .github/workflows/update_apps.yml:61]: Maybe have something a little more descriptive in the message, Like `Automatic update, run by GitHub Actions`.

It'd also be cood if it generatd a list of all apps updated by the workflow!

Comment (theofficialgman) [reply .github/workflows/update_apps.yml:61]: yeah I'm still working on this
adding a check to verify that the new version actually exists for the armhf/arm64 (I've seen a couple of github releases drop support for armhf due to issues and don't want to update the version variable if the file doesn't actually exist)

**Comment by theofficialgman** [comment]
ok formatting aside, is this what you had in mind? https://github.com/Botspot/pi-apps/pull/1330/commits/99306efd4c3ab2c9e5e705bf18e8823c4671bcf5

the full logs are viewable in the github actions under "run app update scripts"
https://github.com/theofficialgman/pi-apps/runs/4670552834?check_suite_focus=true

**Comment by Crilum** [comment]
Yeah! That's great!

**Comment by Crilum** [comment]
It looks like GitHub Actions is only updating btop arm64 and webcord armhf though...

**Comment by theofficialgman** [comment]
> It looks like GitHub Actions is only updating btop arm64 and webcord armhf though...

no its working fine
I'm manually force pushing my branch each time so those two are out of date so I can run the actions again and make sure it updates

**Comment by Crilum** [comment]
Oh ok!

**Comment by Crilum** [comment]
So now we just need to write `update` scripts for each Github installed app right? And, obviously the PR needs to be merged...

**Comment by theofficialgman** [comment]
well 90% of the update scripts for discord and btop are copy paste
this should probably be improved so it can handle install-32, install-64, and install and then added to the api or inside the update_apps.yaml as a function
```
# echo the versions
status "The latest online version is:  $webVer"
status "The current version in Pi-Apps install-32 is:  $pi_apps_ver_32"
status "The current version in Pi-Apps install-64 is:  $pi_apps_ver_64"
#Test to see if version is current:
if [[ "$pi_apps_ver_32" = "$webVer" ]]; then
    #If the version is current do:
    status_green 'Pi-Apps install-32 version is current!'
else
    #If version is not current do:
	if validate_url "$armhf_url"; then
		status_green "Updating pi-apps $app_name install-32 to: $webVer"
		sed -i "s/version=.*/version=${webVer}/g" install-32
		echo "$app_name-armhf " >> /tmp/updated_apps
	else
		warning "Updating $app_name install-32 had been skipped, the upstream file $armhf_url does NOT exist."
	fi
fi

if [[ "$pi_apps_ver_64" = "$webVer" ]]; then
    #If the version is current do:
    status_green 'Pi-Apps install-64 version is current!'
else
	if validate_url "$arm64_url"; then
		status_green "Updating pi-apps $app_name install-64 to: $webVer"
		sed -i "s/version=.*/version=${webVer}/g" install-64
		echo "$app_name-arm64 " >> /tmp/updated_apps
	else
		warning "Updating $app_name install-64 had been skipped, the upstream file $arm64_url does NOT exist."
	fi
fi
```

**Comment by theofficialgman** [comment]
obviously scripts are welcome to have their own custom update script that doesn't use this function, but this usecase is common enough across pi-apps that we should avoid duplication

**Comment by Crilum** [comment]
We can make the above into an `api` function, and another one for `install` type apps..

**Comment by theofficialgman** [comment]
I don't really feel like it should be in the api, since it is only run by the github actions.
see the above commits ^

**Comment by Crilum** [comment]
Yeah, ok. We still need a script for no arch apps (`install`), which I think should be put in the `update_github_script.sh`, as a function, and then put the other armhf/arm64 script inside another function.. Which could then be sourced and run in the workflow...

**Comment by theofficialgman** [comment]
I went ahead and gathered a list of the current apps in pi-apps that use github releases and have the `install` script
most download an `all.deb` file so I'll plan on making that case so it only checks one URL
if the script needs a second, it can make its own custom update file that doesn't use this function

**Comment by Crilum** [comment]
For any update scripts I make, do you want me to just post them here and let you add them to your PR?

**Comment by cycool29** [comment]
Why not just make the `update_github_script.sh` in every app update script?

My idea is: A GitHub workflow triggers the main script that stores functions needed and preparation of updates(like `chmod`ing scripts), and a `for` loop triggers each app update script.

**Comment by cycool29** [comment]
Or maybe `update_github_script.sh` can be compressed as a function? I really don't like it as a script.

**Comment by Crilum** [comment]
That's kind of what it already is, and kink of what I suggested:

to add the `update_github_script.sh` as a function in `api`, but @theofficialgman didn't like it that way so..

**Comment by theofficialgman** [comment]
botspot previously told me he did not want scripts/functions that weren't intended to be run by the user (like these github actions scripts) to be in the normal pi-apps directories

so thats why the scripts and the sourced files are in the `.github/workflow/updates` folder

**Comment by theofficialgman** [comment]
> Why not just make the update_github_script.sh in every app update script?

thats how it used to be
but thats a lot of extra code duplication for many update scripts that will have those commands exactly the same

**Comment by theofficialgman** [comment]
as for the chmoding and for loop stuff, thats already happening in the update_apps.yml

**Comment by cycool29** [comment]
> as for the chmoding and for loop stuff, thats already happening in the update_apps.yml


Yes, so the workflow file can be the `main script` as I described, but it will look really messy (for me).

**Comment by Crilum** [comment]
> > as for the chmoding and for loop stuff, thats already happening in the update_apps.yml
> 
> Yes, so the workflow file can be the `main script` as I described, but it will look really messy (for me).

Why?

**Comment by cycool29** [comment]
> to add the update_github_script.sh as a function in api, but @theofficialgman didn't like it that way so..


No, I don't mean to add it to pi-apps `api`. Instead, we should have another `api` to store extra functions for GitHub Actions.

**Comment by cycool29** [comment]
> Why?

All codes are in blue.

image

But I like colorized code:

image


That's fine, you can leave them as is.

**Comment by Crilum** [comment]
Can you get colored code in YAML? It's blue 'cause it's a sub command in YAML...

**Comment by Crilum** [comment]
> > to add the update_github_script.sh as a function in api, but @theofficialgman didn't like it that way so..
> 
> No, I don't mean to add it to pi-apps `api`. Instead, we should have another `api` (or in the workflow file) to store extra functions for GitHub Actions.

Yeah, I think that would be better..

**Comment by cycool29** [comment]
> Can you get colored code in YAML? It's blue 'cause it's a sub command in YAML...

No, that's why I prefer to put all the code in a separated `main.sh`, and the subcommand in the workflow file will be as simple as:

```yaml
 - name: Run app update scripts
        run: |
          bash .github/workflows/updates/main.sh
```

**Comment by theofficialgman** [comment]
I wouldn't worry about the .yaml file too much
it won't need modification by script writers, its entire job is to load/provide functions, navigate to the app directory, and run any update files

I'm trying my best not to mess with tons of subcommands to avoid variables getting cleared, needing to be exported, etc
lets get the function of the other update scripts in place and then I can edit the main yml later

**Comment by Crilum** [comment]
```
#!/bin/bash
webVer=$(get_release subhra74/xdm)
all_url="https://github.com/subhra74/xdm/releases/download/7.2.11/xdm-setup-${w>
pi_apps_ver="$(cat 'install' | grep 'version=' | sed 's/version=//')"

source update_github_script.sh
```
Is this good for an XDM update script?

**Comment by theofficialgman** [comment]
> source update_github_script.sh

the update script runs in the app folder, so you can't just source the update_github_script.sh file
you have to source its actual path `source $GITHUB_WORKSPACE/.github/workflows/update_github_script.sh`

**Comment by theofficialgman** [comment]
also Xtreme Download Manager doesn't have the `version` variable setup in it yet

**Comment by Crilum** [comment]
Yeah, but we can prepare for it..

**Comment by Crilum** [comment]
XDM install code with version var:

```
#!/bin/bash
version=7.2.11
cd $HOME
mkdir -p xdm 
cd xdm

install_packages openjdk-11-jdk ffmpeg || exit 1

wget -O xdm.tar.xz https://github.com/subhra74/xdm/releases/download/7.2.11/xdm-setup-7.2.11.tar.xz || error "Failed to download XDM setup tar file!"
wget https://github.com/ytdl-org/youtube-dl/releases/download/2021.06.06/youtube-dl || error "Failed to download youtube-dl binary!"
tar -xf xdm.tar.xz || error "Failed to extract XDM setup tar file!"
echo "Executing XDM setup script..."
sudo ./install.sh &>/dev/null || error "Failed to execute XDM setup script!"
sudo chmod +x ./youtube-dl
sudo mv -f ./youtube-dl /opt/xdman || error "Failed to copy youtube-dl binary!"

# Use system java
if [ -d $(dirname $(dirname $(sudo update-alternatives --list java | grep 'java-11-openjdk' --color=none))) ]; then
	sudo rm -rf /opt/xdman/jre/*
	sudo cp -r $(dirname $(dirname $(sudo update-alternatives --list java | grep 'java-11-openjdk' --color=none)))/* /opt/xdman/jre/
elif command -v java &>/dev/null; then
	sudo sed -i s=/opt/xdman/jre/bin/java=$(command -v java)=g /opt/xdman/xdman || error "Failed to change to system Java runtime in /opt/xdman/xdman!"
	sudo sed -i s=/opt/xdman/jre/bin/java=$(command -v java)=g /usr/bin/xdman || error "Failed to change to system Java runtime in /usr/bin/xdman!"
else
	error "OpenJDK-11 installation not found!"
fi
# Fix 'Download failed. Failed to append/convert file parts, please check if the drive is full or write protected'
sudo cp -f $(command -v ffmpeg) /opt/xdman || error "Failed to copy ffmpeg binary!"
# Delete unwanted files after installation
rm -rf $HOME/xdm || error "Failed to remove installation leftovers!"
# Reminder for users to install browser extensiom
echo -e "\nYou may want to install the XDM browser extension. Install it from \e[4mXDM Menu -> Tools -> Browser Monitoring\e[0m page."
```

**Comment by theofficialgman** [comment]
@Crilum no you still don't get what its for
you need to edit the use of the number in the wget command as well
```
wget -O xdm.tar.xz https://github.com/subhra74/xdm/releases/download/${version}/xdm-setup-${version}.tar.xz || error "Failed to download XDM setup tar file!"
```

**Comment by Crilum** [comment]
update script for Angry IP Scanner:

```
#!/bin/bash
version=$(get_release angryip/ipscan)
all_url="https://github.com/angryip/ipscan/releases/download/${version}/xdm-setup-${version}.tar.xz"
pi_apps_ver="$(cat 'install' | grep 'version=' | sed 's/version=//')"

source $GITHUB_WORKSPACE/.github/workflows/update_github_script.sh
```

Is this better, @theofficialgman

**Comment by Crilum** [comment]
> @Crilum no you still don't get what its for you need to remove the hardcoded number in the wget command as well like this
> 
> ```
> wget -O xdm.tar.xz https://github.com/subhra74/xdm/releases/download/${version}/xdm-setup-${version}.tar.xz || error "Failed to download XDM setup tar file!"
> ```

Sorry, I do realize what the var is for, but I was exited and forgot.. :P

**Comment by theofficialgman** [comment]
> update script for Angry IP Scanner:
> 
> ```
> #!/bin/bash
> version=$(get_release angryip/ipscan)
> all_url="https://github.com/angryip/ipscan/releases/download/${version}/xdm-setup-${version}.tar.xz"
> pi_apps_ver="$(cat 'install' | grep 'version=' | sed 's/version=//')"
> 
> source $GITHUB_WORKSPACE/.github/workflows/update_github_script.sh
> ```
> 
> Is this better, @theofficialgman

yeah thats good
the install itself needs editing as well though

**Comment by theofficialgman** [comment]
wait actually no you messed up there

**Comment by theofficialgman** [comment]
https://github.com/Botspot/pi-apps/blob/609c8f7e3de8bc08eb829c8dd08df736639234c4/.github/workflows/update_github_script.sh
refer to the documentation here
don't go changing the variable names
it has to be webVer otherwise it won't work
(I'll probably add a check for that to skip if the variables aren't present)

**Comment by Crilum** [comment]
@theofficialgman look again.. I fixed it before you commented..

**Comment by Crilum** [comment]
How are we going to deal with two GitHub downloads in the same script? See XDM for an example a script like that..

**Comment by theofficialgman** [comment]
> How are we going to deal with two GitHub downloads in the same script? See XDM for an example of a script like that..

there are a couple of options. for one, XDM doesn't necessarily need for youtube-dl to be updated at the same time as xdm itself.

but the update_github_script.sh script could be made more robust to allow for an arbitrary number of supplied URLS.
from the script writers end, it would probably be something like 
```
all_url=( "https://github.com/subhra74/xdm/releases/download/${webVer[0]}/xdm-setup-${webVer[0]}.tar.xz"  "https://github.com/ytdl-org/youtube-dl/releases/download/${webVer[1]}/youtube-dl" )
```

**Comment by cycool29** [comment]
I transferred the `update_github_script.sh` to function:

```sh
update_github_script() {
    local install_32_webVer
	local install_64_webVer
	local install_webVer
	local app
	
    while [[ "${1}" ]]; do
        case "${1}" in
            --app)
                app=${2}
                shift
                ;;
            --install-webVer)
                install_webVer=${2}
                shift
                ;;
            --install-32-webVer)
                install_32_webVer=${2}
                shift
                ;;
            --install-64-webVer)
                install_32_webVer=${2}
                shift
                ;;
            *)
				error "Unknown parameter: ${1}" 
				;;
        esac
		shift
    done
    
    if [ -n "$install_webVer" ] && [ -a "$DIRECTORY/apps/$app/install" ]; then
		install_currentVer="$(grep "version=" "$DIRECTORY/apps/$app/install" | sed s/.*=//g | xargs)"
		if [ "$install_webVer" != "$install_currentVer" ]; then
			echo "Updating $install_currentVer to $install_webVer in $DIRECTORY/apps/$app/install"
			sed -i "s/$(grep "version=" "$DIRECTORY/apps/$app/install")/version=$install_webVer/" "$DIRECTORY/apps/$app/install"
		else
			echo "$DIRECTORY/apps/$app/install version variable match with the web version variable."
		fi
	elif [ -n "$install_32_webVer" ] && [ -a "$DIRECTORY/apps/$app/install-32" ]; then
		install_32_currentVer="$(grep "version=" "$DIRECTORY/apps/$app/install-32" | sed s/.*=//g | xargs)"
		if [ "$install_32_webVer" != "$install_32_currentVer" ]; then
			echo "Updating $install_32_currentVer to $install_32_webVer in $DIRECTORY/apps/$app/install-32"
			sed -i "s/$(grep "version=" "$DIRECTORY/apps/$app/install-32")/version=$install_32_webVer/" "$DIRECTORY/apps/$app/install-32"
		else
			echo "$DIRECTORY/apps/$app/install-32 version variable match with the web version variable."
		fi
	fi
	
    if [ -n "$install_64_webVer" ] && [ -a "$DIRECTORY/apps/$app/install-64" ]; then
		install_64_currentVer="$(grep "version=" "$DIRECTORY/apps/$app/install-64" | sed s/.*=//g | xargs)"
		if [ "$install_64_webVer" != "$install_64_currentVer" ]; then
			echo "Updating $install_64_currentVer to $install_webVer in $DIRECTORY/apps/$app/install-64"
			sed -i "s/$(grep "version=" "$DIRECTORY/apps/$app/install-64")/version=$install_64_webVer/" "$DIRECTORY/apps/$app/install-64"
		else
			echo "$DIRECTORY/apps/$app/install-64 version variable match with the web version variable."
		fi
	fi
}
```

**Comment by Crilum** [comment]
@cycool29 `install_webVer`, `install_32_webVer`, and `install_64_webVer` are supposed to be: `pi_apps_ver`, `pi_apps_ver_32`, and `pi_apps_ver_64`, I think..

**Comment by cycool29** [comment]
> Pretty sure there's only one `webVer`.

Just to avoid if the 32-bit version value is different from the 64-bit version value.

**Comment by Crilum** [comment]
that's a good idea, but we'll have to change the individual update scripts..

**Comment by Crilum** [comment]
Wouldn't you want something more concrete then `sed s/.*=//g`, in case there are more variables then one? Like `sed 's/version=//'`

**Comment by cycool29** [comment]
> Wouldn't you want something more concrete then `sed s/.*=//g`, in case there are more variables then one? Like `sed 's/version=//'`

Look again... I `grep` before I `sed`:
`grep "version=" "$DIRECTORY/apps/$app/install-64" | sed s/.*=//g`

**Comment by theofficialgman** [comment]
> > Pretty sure there's only one `webVer`.
> 
> Just to avoid if the 32-bit version value is different from the 64-bit version value.

we can't have multiple webVer, that would require a complex github api command
the script already exists if there is no version for armhf/arm64 for a specified webVer

**Comment by theofficialgman** [comment]
the if and elif should not exist like that, that is wrong

**Comment by theofficialgman** [comment]
also `--app "Xtreme Download Manager"` this is redundant info if you plan on executing this function from a file already named like the app, which is necessary since there will inevitably be scripts which need custom update files

I guess we could have both... one master update.sh file which just contains a bunch of `update_script_version` function calls, and then other scripts for apps that need custom update scripts that can't be executed by this function

**Comment by theofficialgman** [comment]
also you removed the validate_url part in your function.... this is vital to avoid pushing broken links to the pi-apps repo

**Comment by theofficialgman** [comment]
overall I'm not a fan of the function version, I don't really see any benefit to it and it just makes doing what I said here harder https://github.com/Botspot/pi-apps/pull/1330#issuecomment-1003236359

**Comment by cycool29** [comment]
> we can't actually populate multiple webVer, that would require a complex github api command to find an old release if the latest release does not support one of the archs
the script already exits if there is no version for armhf/arm64 for a specified webVer

That function is not only for the GitHub release. It can be for Sourceforge, GitLab, or even [packages.debian.org](https://packages.debian.org). What this function does is check if the value same as the provided args. If not, change it.

> the if and elif should not exist like that, that is unnecessary, you are already checking for the install file

Do you mean to put the 64-bit `if` under 32-bit `if`?

> also --app "Xtreme Download Manager" this is redundant info if you plan on executing this function from a file already named like the app, which is necessary since there will inevitably be scripts which need custom update files

Then it could be `--app $(basename "$0" | sed s/.sh//g)`.

> also you removed the validate_url part in your function.... this is vital to avoid pushing broken links to the pi-apps repo
(like if armhf or arm64 support gets removed from a new github release, which has happened multiple times)

That will need to extract the URL from the script... I have no idea how to do it.

**Comment by theofficialgman** [comment]
> That function is not only for the GitHub release. It can be for Sourceforge, GitLab, or even packages.debian.org. What this function does is check if the value same as the provided args. If not, change it.

my code also works on more than github releases, its just what I named the file

> That will need to extract the URL from the script... I have no idea how to do it.

yeah its not feasible to extract the url unless you want to modify all scripts to set their URLs in variables at the top of the script (which could be done). Thats why I supplied the user in the update file

**Comment by cycool29** [comment]
> overall I'm not a fan of the function version, I don't really see any benefit to it and it just makes doing what I said here harder #1330 (comment)

Sure, then you can use the script version if that is more comfortable for you. I just don't want to make the process messy. (Script A trigger script B, script B trigger script C...)

**Comment by theofficialgman** [comment]
> but the update_github_script.sh script could be made more robust to allow for an arbitrary number of supplied URLS.

this is the main concern with using a function based approach, I do intend to implement the above. passing in all the variables through `--varname varvale` will be a bit of a nightmare to handle in bash

**Comment by Crilum** [comment]
Nice!

**Comment by theofficialgman** [comment]
there is a bug
I forgot to rename `variable` to `variable2` in the script, 😆

**Comment by theofficialgman** [comment]
alright there we go
that will work assuming the user names their version variables like so:
```
version=
version2=
version3=
```
I could have an override as another input with an array of values incase those default variable names are not desired:
something like
`variable_names=( version_xdm version_ytdl )`

**Comment by Crilum** [comment]
Those are fine. :+1:

**Comment by theofficialgman** [comment]
no changes, just rewrote the commit history so it was compartmentalized and easier to follow

**Comment by theofficialgman** [comment]
ok at this point I've got all the low hanging fruit.
the biggest app that downloads files that could be still use the github actions update mechanism is VSCodium but that will require a custom update script because its filename changes each release

File: .github/workflows/update_apps.yml
Comment (Itai-Nelken) [inline .github/workflows/update_apps.yml:36]:
```suggestion
            if [[ $(wget -S --spider $1  2>&1 | grep 'HTTP/1.1 200 OK'$) ]]; then
```

Comment (cycool29) [inline .github/workflows/update_apps.yml:36]:
```suggestion
            if [ "$(curl -s -o /dev/null -I -w "%{http_code}" $1)" == "200"  ]; then
```

Smarter way.

Comment (theofficialgman) [reply .github/workflows/update_apps.yml:36]: using curl changes the http response codes, its no longer 200

I think the function is overkill anyway, this should work
```
if wget -q --spider "$1"; then
```

**Comment by Crilum** [comment]
When/if DDnet and Droidcam get added, those use GitHub to download stuff, so we'll need update scripts, and the version var in each install script..

**Comment by theofficialgman** [comment]
I skipped over a lot of scripts that technically could be added. Some haven't had github updates in many months, and others just have a really small userbase in pi-apps.
I figured the original script writers can just make PRs in the future to support other apps

**Comment by theofficialgman** [comment]
@Itai-Nelken changed so that it makes a PR instead of pushing to the repo
see my test run of it here: https://github.com/theofficialgman/pi-apps/runs/4744773460?check_suite_focus=true

and the PR it made is here: https://github.com/theofficialgman/pi-apps/pull/1

(still need to change the name so its more useful for us)

**Comment by theofficialgman** [comment]
marking this script as ready for review

script writers and pi-apps maintainers are free to write as complicated update checking scripts as they wish, just make sure to not write files inside the $GITHUB_WORKSPACE directory unless those files/changes are intended to be added to the automated PR

the full ubuntu container is available, you should be able to do apt installs, do test builds, etc if desired

File: .github/workflows/update_apps.yml
Comment (Itai-Nelken) [inline .github/workflows/update_apps.yml:86]: would it be possible to make this a list?
for example:
```markdown
* app1
* app2
```

Comment (theofficialgman) [reply .github/workflows/update_apps.yml:86]: yes, I actually store it as a list in a file, then print it as one line, that can be changed

Comment (theofficialgman) [reply .github/workflows/update_apps.yml:86]: done with https://github.com/Botspot/pi-apps/pull/1330/commits/a30cf499989c797383d7da2da112c0ad2208a866

**Review by Itai-Nelken — APPROVED** [review]
Looks good to me.
What do you think @Botspot?

**Comment by theofficialgman** [comment]
removed the delete branch after merging commit... it doesn't seem to do anything so no purpose in adding it
the delete button exists right after clicking merge, just use that to remove the branch if you don't want it in the github

**Comment by theofficialgman** [comment]
@Botspot is there anything holding you back from merging this?

**Comment by theofficialgman** [comment]
alright cool, now we just wait around till 21:00 UTC time or run it manually to see what it updates

**Comment by Itai-Nelken** [comment]
I have manually run it.

**Pull request merged**
